### PR TITLE
Js doc comment templates

### PR DESCRIPTION
**Pull request by aozgaa**

### CODE REVIEW

File: src/services/utilities.ts [hunk 1]
```diff
@@ -414,6 +414,55 @@ namespace ts {
...
         }
     }
+    
+    export function isInString(sourceFile: SourceFile, position: number) {
+        let token = getTokenAtPosition(sourceFile, position);
+        return token && token.kind === SyntaxKind.StringLiteral && position > token.getStart();
+    }
+
+    export function isInComment(sourceFile: SourceFile, position: number) {
+        return isInCommentHelper(sourceFile, position, c => true);
```
Comment (DanielRosenwasser) [inline src/services/utilities.ts]: Add `/*extraCheck*/`

Comment (aozgaa) [inline src/services/utilities.ts]: Fixed.

**Comment by aozgaa** [comment]
Any thoughts, @CyrusNajmabadi ?

File: src/compiler/scanner.ts [hunk 2]
```diff
@@ -528,13 +535,17 @@ namespace ts {
...
-    // and the next token are returned.If trailing is true, comments occurring between the given
-    // position and the next line break are returned.The return value is an array containing a
-    // TextRange for each comment. Single-line comment ranges include the beginning '//' characters
-    // but not the ending line break. Multi - line comment ranges include the beginning '/* and
-    // ending '*/' characters.The return value is undefined if no comments were found.
+    /**
+     * Extract comments from the given source text starting at the given position. 
+     * The return value is an array containing a TextRange for each comment.
+     * Single-line comment ranges include the beginning '//' characters but not the ending line break.
+     * Multi - line comment ranges include the beginning '/* and ending '<asterick>/' characters.
```
Comment (CyrusNajmabadi) [inline src/compiler/scanner.ts]: ```<asterisk>```?

File: src/services/services.ts [hunk 3]
```diff
@@ -1087,6 +1089,11 @@ namespace ts {
         newText: string;
     }
 
+    export interface TextInsertion {
+        newText: string;
+        cursorOffset: number;
```
Comment (CyrusNajmabadi) [inline src/services/services.ts]: caretOffset

File: src/services/services.ts [hunk 4]
```diff
@@ -6772,6 +6762,67 @@ namespace ts {
             return [];
         }
 
+        /**
+         * Checks if position points to a valid position to add JSDoc comments, and if so,
+         * returns the appropriate scaffolding. Otherwise returns an empty string.
```
Comment (CyrusNajmabadi) [inline src/services/services.ts]: i'm really hating the name 'scaffolding'.

File: src/services/services.ts [hunk 5]
```diff
@@ -6772,6 +6762,67 @@ namespace ts {
...
         }
 
+        /**
+         * Checks if position points to a valid position to add JSDoc comments, and if so,
+         * returns the appropriate scaffolding. Otherwise returns an empty string.
+         * Valid positions are
+         * * outside of comments, statements, and expressions, and
+         * * preceding a function declaration.
+         * 
+         * In VS, we additionally check that:
```
Comment (CyrusNajmabadi) [inline src/services/services.ts]: "Hosts should ideally check that:"

File: src/services/services.ts [hunk 6]
```diff
@@ -6772,6 +6762,67 @@ namespace ts {
...
+         * * The line is all whitespace up to 'position' before performing the insertion.
+         * * If the keystroke sequence "/\*\*" induced the call, we also check that the next
+         * non-whitespace character is '*', which (approximately) indicates whether we added 
+         * the second '*' to complete an existing (JSDoc) comment.
+         * @param fileName The file in which to perform the check.
+         * @param position The (character-indexed) position in the file where the check should
+         * be performed.
+         */
+        function getDocCommentScaffoldingAtPosition(fileName: string, position: number): TextInsertion {
+            // Indicates the position is not appropriate to insert a comment (eg: a string or a regex).
```
Comment (CyrusNajmabadi) [inline src/services/services.ts]: Return 'undefined' for this.  Like most other features.

File: src/services/services.ts [hunk 7]
```diff
@@ -6772,6 +6762,67 @@ namespace ts {
...
+            let start = new Date().getTime();
+            let sourceFile = syntaxTreeCache.getCurrentSourceFile(fileName);
+            log("getDocCommentScaffoldingAtPosition: getCurrentSourceFile: " + (new Date().getTime() - start));
+
+            // Check if in a context where we don't want to perform any insertion
+            if (isInString(sourceFile, position) || isInComment(sourceFile, position) || hasDocComment(sourceFile, position)) {
+                return nullResult;
+            }
+
+            let nodeAtPos = getTokenAtPosition(sourceFile, position);
```
Comment (CyrusNajmabadi) [inline src/services/services.ts]: tokenAtPos instead?

File: src/services/services.ts [hunk 8]
```diff
@@ -6772,6 +6762,67 @@ namespace ts {
...
+            let nodeStart = nodeAtPos.getStart()
+            if (!nodeAtPos || nodeStart < position) {
+                return nullResult;
+            }
+
+            let containingFunction = <FunctionDeclaration>getAncestor(nodeAtPos, SyntaxKind.FunctionDeclaration);
```
Comment (CyrusNajmabadi) [inline src/services/services.ts]: Only functions?  What about methods, accessors, etc.?

File: src/services/services.ts [hunk 9]
```diff
@@ -6772,6 +6762,67 @@ namespace ts {
...
+            }
+
+            let parameters = containingFunction.parameters;
+            let posLineAndChar = sourceFile.getLineAndCharacterOfPosition(position);
+            let lineStart = sourceFile.getLineStarts()[posLineAndChar.line];
+
+            let indentationStr = sourceFile.text.substr(lineStart, posLineAndChar.character).match(/\s*/).toString();
+
+            let docParams = parameters.map((p, index) =>
+                indentationStr + " * @param " + (p.name.kind === SyntaxKind.Identifier ? (<Identifier>p.name).text : "param" + index.toString()) + "\n");
```
Comment (CyrusNajmabadi) [inline src/services/services.ts]: don't hardcode in \n.  Use the host's newline char.

File: src/services/services.ts [hunk 10]
```diff
@@ -6772,6 +6762,67 @@ namespace ts {
...
+
+            let result = 
+                /* opening comment */               "/**\n" +
```
Comment (CyrusNajmabadi) [inline src/services/services.ts]: Inline comments make this harder to read.  Also the formatter will just rip this all up the moment this function ever gets formatted.

File: src/services/services.ts [hunk 11]
```diff
@@ -6772,6 +6762,67 @@ namespace ts {
...
+            let result = 
+                /* opening comment */               "/**\n" + 
+                /* first line for function info */  indentationStr + " * \n" + 
+                /* paramters */                     docParams.reduce((prev, cur) => prev + cur, "") +
+                /* closing comment */               indentationStr + " */" +
+                /* newline if at decl start */      (nodeStart === position ? "\n" + indentationStr : "");
+            
+            let cursorOffset = /* "/**\n" */ 4 + indentationStr.length + /* " * " */ 3;
+
+            return {newText: result, cursorOffset: cursorOffset };
```
Comment (CyrusNajmabadi) [inline src/services/services.ts]: space after open curly.

File: src/services/utilities.ts [hunk 12]
```diff
@@ -414,6 +414,55 @@ namespace ts {
...
+    export function isInComment(sourceFile: SourceFile, position: number) {
+        return isInCommentHelper(sourceFile, position, /*extraCheck*/ c => true);
+    }
+
+    /**
+     * Returns true if the cursor at position in sourceFile is within a comment that additionally
+     * satisfies extraCheck, and false otherwise.
+     */
+    export function isInCommentHelper(sourceFile: SourceFile, position: number, 
+        extraCheck: (c: CommentRange) => boolean): boolean {
```
Comment (CyrusNajmabadi) [inline src/services/utilities.ts]: "predicate"

File: src/services/utilities.ts [hunk 13]
```diff
@@ -414,6 +414,55 @@ namespace ts {
...
+
+    export function isInComment(sourceFile: SourceFile, position: number) {
+        return isInCommentHelper(sourceFile, position, /*extraCheck*/ c => true);
+    }
+
+    /**
+     * Returns true if the cursor at position in sourceFile is within a comment that additionally
+     * satisfies extraCheck, and false otherwise.
+     */
+    export function isInCommentHelper(sourceFile: SourceFile, position: number,
```
Comment (CyrusNajmabadi) [inline src/services/utilities.ts]: wrap and align parameters.

File: src/services/utilities.ts [hunk 14]
```diff
@@ -414,6 +414,55 @@ namespace ts {
...
+     */
+    export function isInCommentHelper(sourceFile: SourceFile, position: number, 
+        extraCheck: (c: CommentRange) => boolean): boolean {
+        let token = getTokenAtPosition(sourceFile, position);
+
+        if (token && position <= token.getStart()) {
+            let commentRanges = getLeadingCommentRanges(sourceFile.text, token.pos);
+
+            return forEach(commentRanges, c => c.pos < position &&
+                // The end marker of a single-line comment does not include the newline character.
```
Comment (CyrusNajmabadi) [inline src/services/utilities.ts]: please pull the comment out of the expression.

File: src/services/utilities.ts [hunk 15]
```diff
@@ -414,6 +414,55 @@ namespace ts {
...
+
+    export function hasDocComment(sourceFile: SourceFile, position: number) {
+        let token = getTokenAtPosition(sourceFile, position);
+
+        let JSDocPrefixRegex = /^\/\*\*\s*/;
+
+        // First, we have to see if this position actually landed in a comment.
+        let commentRanges = getLeadingCommentRanges(sourceFile.text, token.pos);
+
+        return forEach(commentRanges, c => JSDocPrefixRegex.test(sourceFile.text.substring(c.pos, c.end)));
```
Comment (CyrusNajmabadi) [inline src/services/utilities.ts]: You're making a substring on each comment.  Don't do this.  Just literally test the three characters at c.pos for ```/**```

**Comment by DanielRosenwasser** [comment]
Is "scaffolding" an official name for this? Where is it documented? Its name is very vague.

File: src/services/services.ts (hunk 3 above)
Comment (aozgaa) [inline src/services/services.ts]: Do you think renaming the variable to `offsetInNewText` is better?

File: src/services/services.ts (hunk 4 above)
Comment (aozgaa) [inline src/services/services.ts]: Do any of these alternatives work?
* frame
* skeleton
* outline (already used for another feature, we should avoid a conflict)
* blueprint
* sketch
* basis/base
* shell
* mold
* foundation

Of these, I think "skeleton" is best. What do you think @DanielRosenwasser?

File: src/services/services.ts (hunk 5 above)
Comment (aozgaa) [inline src/services/services.ts]: Changed.

File: src/services/services.ts (hunk 6 above)
Comment (aozgaa) [inline src/services/services.ts]: changed.

File: src/services/services.ts (hunk 7 above)
Comment (aozgaa) [inline src/services/services.ts]: changed.

File: src/services/services.ts (hunk 8 above)
Comment (aozgaa) [inline src/services/services.ts]: Just starting with functions. I'll extend it in a separate PR.

File: src/services/services.ts (hunk 4 above)
Comment (RyanCavanaugh) [inline src/services/services.ts]: template?

File: src/services/services.ts (hunk 9 above)
Comment (aozgaa) [inline src/services/services.ts]: changed.

File: src/services/services.ts (hunk 4 above)
Comment (DanielRosenwasser) [inline src/services/services.ts]: Template seems appropriate.

Comment (aozgaa) [inline src/services/services.ts]: Okay.

File: src/services/services.ts (hunk 3 above)
Comment (aozgaa) [inline src/services/services.ts]: I did the above renaming.

File: src/services/services.ts (hunk 10 above)
Comment (aozgaa) [inline src/services/services.ts]: Changed -- consolidated the comments above the code. Though I disagree about the readability, a formatter would wreck it.

File: src/services/services.ts (hunk 11 above)
Comment (aozgaa) [inline src/services/services.ts]: Changed.

File: src/services/services.ts [hunk 16]
```diff
@@ -1086,6 +1088,11 @@ namespace ts {
         newText: string;
     }
 
+    export interface TextInsertion {
+        newText: string;
+        offsetInNewText: number;
```
Comment (CyrusNajmabadi) [inline src/services/services.ts]: offset of what?  "caretOffset" seems liek a reasonable name.  With a comment explaining that it's the offset in 'newText' where the caret shoudl be placed.

File: src/services/services.ts [hunk 17]
```diff
@@ -6804,6 +6794,67 @@ namespace ts {
...
+         * non-whitespace character is '*', which (approximately) indicates whether we added 
+         * the second '*' to complete an existing (JSDoc) comment.
+         * @param fileName The file in which to perform the check.
+         * @param position The (character-indexed) position in the file where the check should
+         * be performed.
+         */
+        function getDocCommentTemplateAtPosition(fileName: string, position: number): TextInsertion {
+            let start = new Date().getTime();
+            let sourceFile = syntaxTreeCache.getCurrentSourceFile(fileName);
+            log("getDocCommentTemplateAtPosition: getCurrentSourceFile: " + (new Date().getTime() - start));
```
Comment (CyrusNajmabadi) [inline src/services/services.ts]: This seems excessive.  Is there a reason we're logging this?  If this is an important metric, maybe we should add the logging for this inside of getCurrentSourceFile itself.

File: src/services/services.ts [hunk 18]
```diff
@@ -6804,6 +6794,67 @@ namespace ts {
...
+                return undefined;
+            }
+
+            let tokenAtPos = getTokenAtPosition(sourceFile, position);
+            let tokenStart = tokenAtPos.getStart()
+            if (!tokenAtPos || tokenStart < position) {
+                return undefined;
+            }
+
+            let containingFunction = <FunctionDeclaration>getAncestor(tokenAtPos, SyntaxKind.FunctionDeclaration);
```
Comment (CyrusNajmabadi) [inline src/services/services.ts]: why only for functions?  What about methods, etc.  What about classes?

File: src/services/services.ts [hunk 19]
```diff
@@ -6804,6 +6794,67 @@ namespace ts {
...
+
+            let indentationStr = sourceFile.text.substr(lineStart, posLineAndChar.character).match(/\s*/).toString();
+
+            const newLine = host.getNewLine ? host.getNewLine() : "\r\n";
+
+            let docParams = parameters.map((p, index) =>
+                indentationStr + " * @param " + (p.name.kind === SyntaxKind.Identifier ? (<Identifier>p.name).text : "param" + index.toString()) + newLine);
+
+            let result = 
+                /* opening comment */               "/**" + newLine +
```
Comment (CyrusNajmabadi) [inline src/services/services.ts]: i really really dislike this.  plus it will be totally destroyed the next time the file is formatted.

File: src/services/services.ts [hunk 20]
```diff
@@ -6804,6 +6794,67 @@ namespace ts {
...
+                indentationStr + " * @param " + (p.name.kind === SyntaxKind.Identifier ? (<Identifier>p.name).text : "param" + index.toString()) + newLine);
+
+            let result = 
+                /* opening comment */               "/**" + newLine + 
+                /* first line for function info */  indentationStr + " * " + newLine + 
+                /* paramters */                     docParams.reduce((prev, cur) => prev + cur, "") +
+                /* closing comment */               indentationStr + " */" +
+                /* newline if at decl start */      (tokenStart === position ? newLine + indentationStr : "");
+            
+            let cursorOffset = /* "/**" */ 3 + /* newLine */ newLine.length + indentationStr.length + /* " * " */ 3;
```
Comment (CyrusNajmabadi) [inline src/services/services.ts]: unreadable comment.  just write:

"/**".length + newLine.length + indentationString.length + " * ".length;

File: src/services/services.ts [hunk 21]
```diff
@@ -6804,6 +6794,67 @@ namespace ts {
...
+
+            return {newText: result, offsetInNewText: cursorOffset };
```
Comment (CyrusNajmabadi) [inline src/services/services.ts]: space after {

**Comment by aozgaa** [comment]
@DanielRosenwasser Changed the name to templates in the code and the PR (but the branch still has the old name).

File: src/services/services.ts (hunk 17 above)
Comment (aozgaa) [inline src/services/services.ts]: Agreed it's excessive. I thought the logging was done around most calls to `getCurrentSourceFile()` but that's incorrect.

File: src/services/services.ts (hunk 18 above)
Comment (aozgaa) [inline src/services/services.ts]: Just starting with functions. I'll extend it in a separate PR.

File: src/compiler/scanner.ts [hunk 22]
```diff
@@ -320,13 +320,20 @@ namespace ts {
 
     /* @internal */
     export function computeLineAndCharacterOfPosition(lineStarts: number[], position: number) {
+        Debug.assert(lineStarts.length && lineStarts[0] === 0, "the first line should start at position 0");
```
Comment (mhegazy) [inline src/compiler/scanner.ts]: can we remove these asserts from here. the server calls this function at least once for every call. we are not removing asserts in retail version and this code has not changed in a relay long time. so i do not see the value.

Comment (aozgaa) [inline src/compiler/scanner.ts]: Removed.

File: src/services/services.ts (hunk 3 above)
Comment (aozgaa) [inline src/services/services.ts]: Changed to `caretOffset`.

File: src/services/services.ts [hunk 23]
```diff
@@ -6804,6 +6795,74 @@ namespace ts {
...
+        /**
+         * Checks if position points to a valid position to add JSDoc comments, and if so,
+         * returns the appropriate template. Otherwise returns an empty string.
+         * Valid positions are
+         * * outside of comments, statements, and expressions, and
+         * * preceding a function declaration.
+         *
+         * Hosts should ideally check that:
+         * * The line is all whitespace up to 'position' before performing the insertion.
+         * * If the keystroke sequence "/\*\*" induced the call, we also check that the next
```
Comment (CyrusNajmabadi) [inline src/services/services.ts]: why the backslashes?

File: src/services/services.ts (hunk 18 above)
Comment (CyrusNajmabadi) [inline src/services/services.ts]: ok.  add a note.

File: src/services/services.ts [hunk 24]
```diff
@@ -6804,6 +6795,74 @@ namespace ts {
...
+                return undefined;
+            }
+
+            let parameters = containingFunction.parameters;
+            let posLineAndChar = sourceFile.getLineAndCharacterOfPosition(position);
+            let lineStart = sourceFile.getLineStarts()[posLineAndChar.line];
+
+            let indentationStr = sourceFile.text.substr(lineStart, posLineAndChar.character).match(/\s*/).toString();
+
+            const newLine = host.getNewLine ? host.getNewLine() : "\r\n";
```
Comment (CyrusNajmabadi) [inline src/services/services.ts]: do we not do this elsewhere?  can we extract this out as a helper for all features to use?

File: src/services/services.ts [hunk 25]
```diff
@@ -6804,6 +6795,74 @@ namespace ts {
...
+
+            let docParams = parameters.map((p, index) =>
+                indentationStr + " * @param " + (p.name.kind === SyntaxKind.Identifier ? (<Identifier>p.name).text : "param" + index.toString()) + newLine);
```
Comment (CyrusNajmabadi) [inline src/services/services.ts]: no need to call ```.toString()```

File: src/services/services.ts [hunk 26]
```diff
@@ -6804,6 +6795,74 @@ namespace ts {
...
+            // * The opening comment line
+            // * the first line (without a param) for the object's untagged info (this is also where the caret ends up)
+            // * the '@param'-tagged lines
+            // * TODO: other tags.
+            // * the closing comment line
+            // * if the caret was directly in front of the object, then we add an extra line and indentation.
+            let result =
+                "/**" + newLine +
+                indentationStr + " * " + newLine +
+                docParams.reduce((prev, cur) => prev + cur, "") +
```
Comment (CyrusNajmabadi) [inline src/services/services.ts]: docParams.join("") is much more efficient.

File: src/services/services.ts [hunk 27]
```diff
@@ -6804,6 +6795,74 @@ namespace ts {
...
+
+            // A doc comment consists of the following
+            // * The opening comment line
+            // * the first line (without a param) for the object's untagged info (this is also where the caret ends up)
+            // * the '@param'-tagged lines
+            // * TODO: other tags.
+            // * the closing comment line
+            // * if the caret was directly in front of the object, then we add an extra line and indentation.
+            let result =
+                "/**" + newLine +
```
Comment (CyrusNajmabadi) [inline src/services/services.ts]: extract this bit:

```
 "/**" + newLine +
+                indentationStr + " * "
```
into a variable called 'preamble'

File: src/services/services.ts [hunk 28]
```diff
@@ -6804,6 +6795,74 @@ namespace ts {
...
+                indentationStr + " * " + newLine +
+                docParams.reduce((prev, cur) => prev + cur, "") +
+                indentationStr + " */" +
+                (tokenStart === position ? newLine + indentationStr : "");
+
+            let cursorOffset = "/**".length + newLine.length + indentationStr.length + " * ".length;
```
Comment (CyrusNajmabadi) [inline src/services/services.ts]: Then you can change this to ```= preamble.length```.

DRY.

File: src/services/utilities.ts [hunk 29]
```diff
@@ -414,6 +414,57 @@ namespace ts {
...
         }
     }
+    
+    export function isInString(sourceFile: SourceFile, position: number) {
+        let token = getTokenAtPosition(sourceFile, position);
+        return token && token.kind === SyntaxKind.StringLiteral && position > token.getStart();
+    }
+
+    export function isInComment(sourceFile: SourceFile, position: number) {
+        return isInCommentHelper(sourceFile, position, /*predicate*/ c => true);
```
Comment (CyrusNajmabadi) [inline src/services/utilities.ts]: I would make the predicate optional.  And if it is not htere, then you don't check for it in the isInCommentHelper code.  This save an alloc here.

File: src/services/utilities.ts [hunk 30]
```diff
@@ -414,6 +414,57 @@ namespace ts {
...
+        return false;
+    }
+
+    export function hasDocComment(sourceFile: SourceFile, position: number) {
+        let token = getTokenAtPosition(sourceFile, position);
+
+        // First, we have to see if this position actually landed in a comment.
+        let commentRanges = getLeadingCommentRanges(sourceFile.text, token.pos);
+
+        return forEach(commentRanges, c => jsDocPrefix);
```
Comment (CyrusNajmabadi) [inline src/services/utilities.ts]: this makes no sense, and i'm surprised it even compiled.  it either needs to be: ```, c => jsDocPrefix(c))``` or just ```, jsDocPrefix)```

**Comment by CyrusNajmabadi** [comment]
With a couple more changes, this will LGTM.

File: src/compiler/types.ts [hunk 31]
```diff
@@ -586,9 +586,9 @@ namespace ts {
      * Several node kinds share function-like features such as a signature,
      * a name, and a body. These nodes should extend FunctionLikeDeclaration.
      * Examples:
-     *  FunctionDeclaration
-     *  MethodDeclaration
-     *  AccessorDeclaration
+     * * FunctionDeclaration
```
Comment (DanielRosenwasser) [inline src/compiler/types.ts]: I'd make these dashes

File: src/services/services.ts (hunk 23 above)
Comment (aozgaa) [inline src/services/services.ts]: The backslashes escape the asterisk when it is rendered by hovering over the definition or a call of `getDocCommentTemplateAtPosition()`.

File: src/harness/fourslash.ts [hunk 32]
```diff
@@ -1941,6 +1941,33 @@ module FourSlash {
...
+            const name = "verifyDocCommentTemplate";
+            let actual = this.languageService.getDocCommentTemplateAtPosition(this.activeFile.fileName, this.currentCaretPosition);
+
+            if (expected === undefined) {
+                if (actual === undefined) {
+                    return;
+                }
+                else {
+                    this.raiseError(name + ' failed - expected no template but got {newText: \"' + actual.newText + '\" caretOffset: ' + actual.caretOffset + '}');
+                }
```
Comment (DanielRosenwasser) [inline src/harness/fourslash.ts]: I might just make this

```TypeScript
if (expected === undefined) {
    if (actual) {
        this.raiseError(name + ' failed - expected no template but got {newText: \"' + actual.newText + '\" caretOffset: ' + actual.caretOffset + '}');
    }

    return;
}

Then remove the `else` block following this

File: src/services/services.ts [hunk 33]
```diff
@@ -1086,6 +1088,12 @@ namespace ts {
         newText: string;
     }
 
+    export interface TextInsertion {
+        newText: string;
+        /** The position in newText the caret should point to after the insertion. */
+        caretOffset: number;
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: It would be interesting if this was actually a range so that the user could just start typing over a default region.

Comment (DanielRosenwasser) [inline src/services/services.ts]: Actually it would potentially be interesting if this was a series of ranges that one could tab through.

File: src/services/services.ts [hunk 34]
```diff
@@ -6768,7 +6759,7 @@ namespace ts {
         function getIndentationAtPosition(fileName: string, position: number, editorOptions: EditorOptions) {
             let start = new Date().getTime();
             let sourceFile = syntaxTreeCache.getCurrentSourceFile(fileName);
-            log("getIndentationAtPosition: getCurrentSourceFile: " + (new Date().getTime() - start));
+            log("getIndentationAtPosition: getCurrentSourceFile: " + (new Date().getTime() - start));
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: Remove trailing space. :smile:

File: src/services/services.ts (hunk 24 above)
Comment (aozgaa) [inline src/services/services.ts]: There is a separate PR out for making the helper (#4133). Once that get's merged I'll just call the helper.

File: src/services/services.ts [hunk 35]
```diff
@@ -6804,6 +6795,74 @@ namespace ts {
...
+
+            if (!containingFunction || containingFunction.getStart() < position) {
+                return undefined;
+            }
+
+            let parameters = containingFunction.parameters;
+            let posLineAndChar = sourceFile.getLineAndCharacterOfPosition(position);
+            let lineStart = sourceFile.getLineStarts()[posLineAndChar.line];
+
+            let indentationStr = sourceFile.text.substr(lineStart, posLineAndChar.character).match(/\s*/).toString();
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: This seems suspect, double check what `toString` does and let's  make it more explicit anyway

File: src/services/services.ts [hunk 36]
```diff
@@ -6804,6 +6795,74 @@ namespace ts {
...
+            let result =
+                "/**" + newLine +
+                indentationStr + " * " + newLine +
+                docParams.reduce((prev, cur) => prev + cur, "") +
+                indentationStr + " */" +
+                (tokenStart === position ? newLine + indentationStr : "");
+
+            let cursorOffset = "/**".length + newLine.length + indentationStr.length + " * ".length;
+
+            return { newText: result, caretOffset: cursorOffset };
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: If you don't end up making this `preamble.length`, I might just rename `cursorOffset` to `caretOffset` so you can use a shorthand property.

File: src/compiler/types.ts (hunk 31 above)
Comment (aozgaa) [inline src/compiler/types.ts]: Looks much better in the popup and above the `FunctionLikeDeclaration`.

File: src/harness/fourslash.ts (hunk 32 above)
Comment (aozgaa) [inline src/harness/fourslash.ts]: The only issue is that we don't want to perform the remaining checks if both `expected === undefined` and `actual === undefined` (we'll get null-dereference errors).

File: src/services/services.ts (hunk 33 above)
Comment (aozgaa) [inline src/services/services.ts]: YAGNI. But when you do need it, by all means change it up.

Comment (DanielRosenwasser) [inline src/services/services.ts]: > YAGNI

YAGNI implies I'm trying to cover us for when/if we adopt that model. I'm suggesting that we just adopt that model. Maybe @CyrusNajmabadi can weigh in.

File: src/harness/fourslash.ts (hunk 32 above)
Comment (DanielRosenwasser) [inline src/harness/fourslash.ts]: Right, but the `return;` will jump out before that happens

File: src/services/services.ts (hunk 33 above)
Comment (aozgaa) [inline src/services/services.ts]: I suppose I just don't see how those functionalities are relevant to this PR or feature. 

Do you want us to emit something like
```ts
/**
 * <+function_description+>
 * @param a <+param_description+>
 * @param b <+param_description+>
 */
```

and then we can iterate between them by tabbing or some hotkey (eg: <C-j>)?

(My understanding of the functionality is inspired by the Latex-Suite plugin for vim)

Comment (aozgaa) [inline src/services/services.ts]: ... Because that seems outside the scope of this PR.

File: src/services/services.ts (hunk 36 above)
Comment (aozgaa) [inline src/services/services.ts]: changed to `preamble.length`.

File: src/harness/fourslash.ts (hunk 32 above)
Comment (aozgaa) [inline src/harness/fourslash.ts]: yup. changed that.

**Comment by aozgaa** [comment]
@JsonFreeman thoughts?

File: src/services/services.ts [hunk 37]
```diff
@@ -6804,6 +6795,78 @@ namespace ts {
...
+
+            if (!containingFunction || containingFunction.getStart() < position) {
+                return undefined;
+            }
+
+            let parameters = containingFunction.parameters;
+            let posLineAndChar = sourceFile.getLineAndCharacterOfPosition(position);
+            let lineStart = sourceFile.getLineStarts()[posLineAndChar.line];
+
+            let indentationStr = sourceFile.text.substr(lineStart, posLineAndChar.character);
```
Comment (JsonFreeman) [inline src/services/services.ts]: Is it possible for this comment operation to be triggered in the same keystroke as formatting? If so, is this the sourceFile's text before or after the formatting edits have taken effect? If it is before, then would indentationStr still be correct?

File: src/services/services.ts [hunk 38]
```diff
@@ -6804,6 +6795,78 @@ namespace ts {
...
+
+            // TODO: call a helper method instead once PR #4133 gets merged in.
+            const newLine = host.getNewLine ? host.getNewLine() : "\r\n";
+
+            let docParams = parameters.map((p, index) =>
```
Comment (JsonFreeman) [inline src/services/services.ts]: Is it worth making an array that will only be joined later?

File: src/services/services.ts (hunk 37 above)
Comment (aozgaa) [inline src/services/services.ts]: Not that I'm aware of.

I think our handling of doc-comment formatting, while distinct from this PR, could be improved. If the comment has the shape of a JSDoc comment, then there appears to be an idiomatic way to format them.

File: src/services/services.ts (hunk 38 above)
Comment (aozgaa) [inline src/services/services.ts]: I think the intent is clearer, but if it saves alloc's I'll eliminate it.

Comment (aozgaa) [inline src/services/services.ts]: changed.

**Comment by aozgaa** [comment]
@CyrusNajmabadi @DanielRosenwasser @paulvanbrenk Can you take a final read?

**Comment by aozgaa** [comment]
@CyrusNajmabadi @DanielRosenwasser @paulvanbrenk could you take a final read of the code?
